### PR TITLE
[BD-05] [TNL-7308] Add method to collect attachments data to OraAggregateData

### PR DESCRIPTION
**Pull request by 0x29a**
This PR adds `collect_ora2_attachments` method, which collects all submission attachments and represents them in a way useful to create attachments zip archive.

**JIRA tickets**:
- [TNL-7308](https://openedx.atlassian.net/browse/TNL-7308)
- [BLENDED-473](https://openedx.atlassian.net/browse/BLENDED-473)

**Dependencies**:
This PR is a dependency for https://github.com/edx/edx-platform/pull/24541.

**Merge deadline**: None

**Testing instructions**:

Since there are just one new method and it's covered with test, I think CI pass is enough to consider this tested. If not, this method can be tested in dependent pull request.

**Reviewers**
- [ ] @xitij2000 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @0x29a! I've created [BLENDED-473](https://openedx.atlassian.net/browse/BLENDED-473) to keep track of it in Jira.  More details are on the [BD-05](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529970851/BD-05+ORA+v3+Usability+Moderation) project page.

When this pull request is ready, tag your edX technical lead.

File: openassessment/data.py [hunk 1]
```diff
@@ -670,3 +672,52 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
...
+                    texts: [...text dicts...]
+                },
+                ...
+            },
+            ...
+        }
+        ```
+        """
+
+        all_submission_information = list(sub_api.get_all_course_submission_information(course_id, 'openassessment'))
```
Comment (xitij2000) [inline openassessment/data.py:695]: Why convert this to a list when it's going to be iterated over normally?

```suggestion
        all_submission_information = sub_api.get_all_course_submission_information(course_id, 'openassessment')
```

File: openassessment/data.py [hunk 2]
```diff
@@ -670,3 +672,52 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
...
+
+        result = {}
+
+        for student_info, submission, _ in all_submission_information:
+            files = result.setdefault(
+                student_info['item_id'], {}
+            ).setdefault(
+                student_info['student_id'], {'attachments': [], 'texts': []}
+            )
```
Comment (xitij2000) [inline openassessment/data.py:704]: This is a bit hard to follow, can you break it down into multiple lines?

File: openassessment/data.py (hunk 1 above)
Comment (0x29a) [reply openassessment/data.py:695]: Ah, that's a copy-paste from another function. No point in turning this into list here.

File: openassessment/data.py [hunk 3]
```diff
@@ -670,3 +676,178 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
...
+                    file_data['type'],
+                    file_data['description'],
+                    file_data['size'],
+                    file_data['file_path'],
+                )
+            )
+
+        output_buffer = StringIO()
+        csvwriter = csv.writer(output_buffer)
+        csvwriter.writerows([header] + rows)
```
Comment (xitij2000) [inline openassessment/data.py:739]: Note: 

If you use DictWriter you can simplify this code a lot. This whole bit a lot. All this can become:
```suggestion
        output_buffer = StringIO()
        csvwriter = csv.DictWriter(output_buffer, raw_fields)
        csvwriter.writerow(header)
        csvwriter.writerows(file_data)
```

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "edx-ora2",
-  "version": "2.8.10",
+  "version": "2.8.11",
```
Comment (xitij2000) [inline package.json:3]: I think this justifies a bump to 2.9.0 since it add a new functionality.

**Comment by 0x29a** [comment]
@davestgermain, this one is ready for review, as well as https://github.com/edx/edx-platform/pull/24541.

File: openassessment/data.py (hunk 3 above)
Comment (davestgermain) [reply openassessment/data.py:739]: I agree that [DictWriter](https://docs.python.org/3/library/csv.html#csv.DictWriter) would be cleaner. Since you're defining the column names in this class (`collect_ora2_submission_files`), you don't have to repeat them here.

File: openassessment/data.py [hunk 5]
```diff
@@ -670,3 +676,178 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
...
+                    'size': len(content),
+                    'file_path': os.path.join(
+                        str(course_id),
+                        student['item_id'],
+                        student['student_id'],
+                        file_name,
+                    )
+                })
+
+        return files
```
Comment (davestgermain) [inline openassessment/data.py:853]: An optimization would be to make this method an iterator instead of returning a list of files. Just `yield` each file dict.

File: openassessment/data.py [hunk 6]
```diff
@@ -670,3 +676,178 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
                 result[item_id][status] += 1
 
         return result
+
+
+class OraDownloadData:
```
Comment (davestgermain) [inline openassessment/data.py:681]: I don't really like that this is a class that only has classmethods. That seems to indicate that these are better off as simple functions in `data.py`. But I see that you're copying the pattern that was here already, in `OraAggregateData`. So I suppose it's ok for now.

File: openassessment/data.py [hunk 7]
```diff
@@ -670,3 +676,178 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
...
+            for file_data in submission_files_data:
+                file_content = (
+                    cls._download_file_by_key(file_data['key'])
+                    if file_data['type'] == cls.ATTACHMENT
+                    else file_data['content']
+                )
+
+                zip_file.writestr(file_data['file_path'], file_content)
+
+            downloads_csv = cls._create_downloads_csv(submission_files_data)
```
Comment (davestgermain) [inline openassessment/data.py:781]: If the download csv is only ever used in this zip file, why not just create the csv as you iterate `submission_files_data`?

**Comment by 0x29a** [comment]
Thanks for suggesions, @davestgermain and @xitij2000.

`collect_ora2_submission_files` now yields results instead of constructing list. Also, I managed to get rid of separate `_create_downloads_csv`, now csv is generated along with writing files in zip.

File: openassessment/data.py (hunk 6 above)
Comment (0x29a) [reply openassessment/data.py:681]: >as simple functions in data.py

I agree. Probably, since there are already two such classes, we can split `data.py` on three different files, like `aggregate.py`, `download.py` and `utils.py` (for `CsvWriter`).

**Comment by 0x29a** [comment]
>Could you just squash the commits?

Done, @davestgermain.
